tests/infra_test.go

```diff
@@ -226,50 +226,53 @@ var _ = Describe("[Serial]Infrastructure", func() {
 			})
 
 			AfterEach(func() {
-				By("removing the taint from the tainted node")
-				var taints []k8sv1.Taint
-
-				for _, taint := range selectedNode.Spec.Taints {
-					if taint.Key != "CriticalAddonsOnly" {
-						taints = append(taints, taint)
+				if selectedNode != nil {
+					By("removing the taint from the tainted node")
+					var taints []k8sv1.Taint
+					for _, taint := range selectedNode.Spec.Taints {
+						if taint.Key != "CriticalAddonsOnly" {
+							taints = append(taints, taint)
+						}
 					}
-				}
 
-				nodeCopy := selectedNode.DeepCopy()
-				nodeCopy.ResourceVersion = ""
+					nodeCopy := selectedNode.DeepCopy()
+					nodeCopy.ResourceVersion = ""
 
-				err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
-					_, err := virtClient.CoreV1().Nodes().Update(nodeCopy)
-					return err
-				})
-				Expect(err).ShouldNot(HaveOccurred())
+					err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+						_, err := virtClient.CoreV1().Nodes().Update(nodeCopy)
+						return err
+					})
+					Expect(err).ShouldNot(HaveOccurred())
+				}
 			})
 
 			It("[test_id:4134] kubevirt components on that node should not evict", func() {
 
 				By("finding all kubevirt pods")
-				pods, err := virtClient.CoreV1().Pods(flags.KubeVirtInstallNamespace).List(metav1.ListOptions{})
+				kubevirtPods, err := virtClient.CoreV1().Pods(flags.KubeVirtInstallNamespace).List(metav1.ListOptions{})
 				Expect(err).ShouldNot(HaveOccurred(), "failed listing kubevirt pods")
-				Expect(len(pods.Items)).To(BeNumerically(">", 0), "no kubevirt pods found")
-
-				By("finding all schedulable nodes")
-				schedulableNodesList := tests.GetAllSchedulableNodes(virtClient)
-				schedulableNodes := map[string]*k8sv1.Node{}
-				for _, node := range schedulableNodesList.Items {
-					schedulableNodes[node.Name] = node.DeepCopy()
-				}
+				Expect(len(kubevirtPods.Items)).To(BeNumerically(">", 0), "no kubevirt pods found")
 
 				By("selecting one compute only node that runs kubevirt components")
 				// master nodes should never have the CriticalAddonsOnly taint because core components might not
 				// tolerate this taint because it is meant to be used on compute nodes only. If we set this taint
 				// on a master node, we risk in breaking the test cluster.
-				for _, pod := range pods.Items {
-					node := schedulableNodes[pod.Spec.NodeName]
-					if _, isMaster := node.Labels["node-role.kubernetes.io/master"]; isMaster {
-						continue
+				Eventually(func() *k8sv1.Node {
+					schedulableNodesList := tests.GetAllSchedulableNodes(virtClient)
```

Comment:
How can we get to the situation we have pods but no schedulable nodes?

Reply:
see here that it happened,
https://github.com/kubevirt/kubevirt/issues/4158
when it runs in check-tests-for-flakes, its the first test that runs on a fresh cluster, and then the cluster seems sometimes to have a little hiccup after it has the pod list, for a short time (it happens on fresh clusters sometimes from what i saw in the past, and it looks like it what happens here).

Reply:
So it seems that we have a bigger problem. We might start running the tests on a not ready cluster.
We cannot add eventually's to each one of the tests waiting for the cluster to be ready.

I think we should understand the root cause of the problem and fix it. Fixing this specific test is confusing and doesn't solve the problem for the other tests that might have the same issue.

Reply:
We didn't see it hurt other tests explicitly.
This test is sensitive for it and it cause flakes on the `check-lane-for-flakes` which runs only the files that are changed in the PR, and then it can be the first test, so i think its fine and good to have this test stable, in case we need we can do a larger scale fix on other PR.

Reply:
The PR can keep the other refactories it has. But I think the use of the `Eventually` is wrong and may hide other problems.
The test shouldn't assume the cluster may have "hiccups". We shouldn't start adding `Eventually`s to our tests to cover those "hiccups".
Can you please explain me what is so special in this test that makes it "especially sensitive"? Should we now fix all the tests the are using `tests.GetAllSchedulableNodes`?

I believe that the issue that you found it bigger than this specific test. We are running the tests too early. And that's what I believe should be fixed.

Reply:
> The PR can keep the other refactories it has. But I think the use of the `Eventually` is wrong and may hide other problems.
> The test shouldn't assume the cluster may have "hiccups". We shouldn't start adding `Eventually`s to our tests to cover those "hiccups".
> Can you please explain me what is so special in this test that makes it "especially sensitive"? Should we now fix all the tests the are using `tests.GetAllSchedulableNodes`?

Its special that when it runs as the first test its more sensitive than others (from what we actually see, not about what we didn't see yet), as it has a flow where it samples the nodes scheduable flag, which other tests doesnt do directly, but implicit on some stages in their life time maybe.

> 
> I believe that the issue that you found it bigger than this specific test. We are running the tests too early. And that's what I believe should be fixed.

It was said implicit in the above text "This test is especially sensitive for it and it cause flakes on the check-lane-for-flakes which runs only the files that are changed in the PR, and then it can be the first test".
But i think we should first see it happens and affect other tests and then act, at least to some degree, or if we simulate it, we also need to prioritize it, i wont suggest to put it highest priority before we dive into it more and see it harms (it might harm here and there already, maybe the compile time for example of all the tests which is longer that just compiling this test, make the transient go
).
I honestly don't think we should block this PR, nor suggested to add 'Eventually' on other PRs, here its suitable, and natural as part of the other changes, in order to reduce noise. 
We already learned from it what we should and can act ortogonal.

Reply:
Sorry, but I'm totally against putting an `Eventually` to cover an error we don't fully understand just to reduce the noise.
Using `Eventually` makes sense when we are waiting for a result after an async update.
Here it is not the case. We are waiting for schedulable nodes on a cluster with running pods!
All our tests assume the cluster is ready, I truly believe it is wrong to soften this assumption on some tests just to reduce noise.

Reply:
I understand, i just think that since we dont see the error affecting other tests, but just this one, and just on the `check-for-flakes` lane, and not even on flake-finder, i found it safe, else i would think the same.